### PR TITLE
feat(core): document layout

### PR DESCRIPTION
**Pull request by hermanwikner**
### Description

This pull request introduces `document.components.layout` to the Components API, enabling the customization of the document layout. The pull request also refactors the comments plugin to utilize this new capability within the Components API. As a result, all comments-specific logic that was previously implemented directly in `desk` is now removed and is instead configured in the comments plugin. That is, the comments feature is now entirely implemented as a plugin, utilizing existing APIs.

```ts
import {defineConfig} from 'sanity'
import {ContextProvider} from './ContextProvider'

export default defineConfig({
  // ...

  document: {
    components: {
      unstable_layout: (props) => {
        const {documentId, documentType} = props

        return (
          <ContextProvider documentId={documentId} documentType={documentType}>
            {props.renderDefault(props)}
          </ContextProvider>
        )
      },
    },
  },
})
```

**Things to keep in mind:**
- Unlike other components in the Components API, which are resolved in `core` and made available universally across all tools, the `document.components.layout` requires resolution within each individual tool. This is because there is no predefined concept of a specific document layout. That is, it is the responsibility of each individual tool to resolve the document layout component from the Components API middleware chain. To accomplish this, the `useMiddlewareComponents` hook can be used, which is exported from the `sanity` package ([see example here in `desk`](https://github.com/sanity-io/sanity/blob/422af38012693bebd662a3738f368958e7d59925/packages/sanity/src/desk/panes/document/document-layout/useDocumentLayoutComponent.ts)).
- Although the `document.components.layout` component receives the `documentId` and `documentType` in props, modifying these props passed to `renderDefault` will have no effect. This is because the `documentId` and `documentType` used in `desk` are resolved from the URL. That is, these props are currently being used as ‘read-only’ props in order to know which context you’re in.
-  In the context of `desk`, which uses an advanced pane system layout, customizing the UI aspects of the document layout might lead to unexpected results and is not recommended. Instead, the API is best used as a way to add context providers to the document layout.

### What to review

- Everything should work as usual
- Make sure that customizations made to `document.components.layout` are properly applied
- Removing the `comments` plugin from `deskTool` should remove the feature entirely and no errors should appear

### Notes for release

N/A

### CODE REVIEW

**Review by ricokahler — COMMENTED** [review]
I really like this API. It follows the same conventions for the `studio.components` API and enables a lot of cool plugins. I think this API could pair well with the changes I proposed in https://github.com/sanity-io/sanity/pull/4938 a while back.

I'd like to I think about refactoring the component resolution of all component API into the config resolver for consistency's sake but that's a different thing. This is a very solid API addition! Looking forward to discussing this in our Studio DX sync tomorrow!

File: packages/sanity/src/desk/comments/plugin/document-layout/CommentsDocumentLayout.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    openInspector(COMMENTS_INSPECTOR_NAME)
+  }, [inspector?.name, openInspector])
+
+  // If comments are not enabled, render the default document layout
+  if (!commentsEnabled) {
+    return props.renderDefault(props)
+  }
+
+  return (
+    <CommentsEnabledProvider documentId={documentId} documentType={documentType}>
```
[the remark this answers is not in the document]
Comment (hermanwikner) [reply packages/sanity/src/desk/comments/plugin/document-layout/CommentsDocumentLayout.tsx:39]: Yes! 👍

File: packages/sanity/src/desk/comments/src/hooks/useResolveCommentsEnabled.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import {useMemo} from 'react'
+import {getPublishedId, useFeatureEnabled, useSource} from 'sanity'
+
+/**
+ * @internal
+ * A hook that resolves if comments are enabled for the current document and document type
+ * and if the feature is enabled for the current project.
+ */
+export function useResolveCommentsEnabled(documentId: string, documentType: string): boolean {
```
[the remark this answers is not in the document]
Comment (hermanwikner) [reply packages/sanity/src/desk/comments/src/hooks/useResolveCommentsEnabled.ts:9]: Yes, but there's already a hook called `useCommentsEnabled` that returns the context value of `CommentsEnabledContext`. The idea with this hook is that you can check if comments are enabled when you're outside of that context!

File: packages/sanity/src/desk/panes/document/document-layout/DocumentLayout.tsx [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    (path: Path) => {
+      onPathOpen(path)
+      onFocus(path)
+    },
+    [onPathOpen, onFocus],
+  )
+
+  if (!schemaType) {
+    return (
+      <DocumentLayoutError
```
Comment (ricokahler) [inline packages/sanity/src/desk/panes/document/document-layout/DocumentLayout.tsx:145]: this is a nice refactor!

**Review by ricokahler — APPROVED** [review]
very nice! love the refactors you did too!

File: packages/sanity/src/core/config/components/useMiddlewareComponents.ts [hunk 4]
```diff
@@ -34,7 +34,27 @@ function _createMiddlewareComponent<T extends {}>(
   }
 }
 
-/** @internal */
+/**
+ * @internal
```
Comment (bjoerge) [inline packages/sanity/src/core/config/components/useMiddlewareComponents.ts:38]: Thanks for documenting this <3

**Review by bjoerge — APPROVED** [review]
Excellent work! Thanks for taking the time to add documentation and tests as well.

**Pull request merged**